### ISSUE TITLE
Test all debug options and split profiling

### ISSUE THREAD

**Issue by alerque** [sile-typesetter/sile#1635]
Currently we aren't testing all the debug options, mainly just the `versions` flag when building the manual. See [this comment](https://github.com/sile-typesetter/sile/pull/1634#issuecomment-1330352142) for a suggestion to fix this and #1626 for why it is something that should be tested.

Unfortunately running *all* the debug flags is prohibitively slow, something I don't really want to add to our already somewhat tardy CI runners. Additionally it would currently expect a profiling library to be available.

Lets think about a way we can test this regularly without slowing things down too much. Also splitting profiling into a separate option instead of a special case of debug flags might be in order.

**Comment by raphCode** [comment]
> Unfortunately running _all_ the debug flags is prohibitively slow

Would a minimal test document that touches all the debug functions suffice? The idea would be to only run this single document with `--debug all`, not slowing down other tests.
If this would still be too big / slow, maybe several minimal documents each using a selection of debug options?

**Comment by alerque** [comment]
Yes, such a document *could be* conceived. But it it's a matter of just making up a debug statement of each type, you have to invent scenarios where all the existing debug conditions might get bumped into, and there are lots of them at weird points. It would be quite a chore to come up with a document that invented a scenario to trigger each possible debug statement. The advantage of the manual is that it already *is* a kind of stress tests that demonstrates almost every package and basic scenario, frequently with the most troublesome combinations.

Another thought is that we could tweak the `SU.debug()` function to protect it with `pcall()` so that debugging can't break document rendering, then manually test the manual once and move on. Unfortunately that may work only with my idea of using debug functions (not strings) not precomposted strings. The current mix and match will still error early in many cases. I've already started that conversion (because it also makes debug statements faster when not used) but it isn't finished.

**Comment by raphCode** [comment]
> Another thought is that we could tweak the `SU.debug()` function to protect it with `pcall()`
> Unfortunately that may work only with my idea of using debug functions (not strings) not precomposted strings.



If you don't mind making a kinda hacky solution, I believe it should be possible to use `m4` to wrap all `SU.debug()` calls so that the string concatenation happens inside a function protected by `pcall()`.
To make this solution a bit less hacky, show macro usage explicit by making it look less like a Lua function call and more like a C macro or something:
`SU_DEBUG("typesetter", "this may fail" .. dangerous_nil_value_here)`

---

I wanted to share more thoughts on this, even if they are likely not feasible :
- typed code / static analysis could catch errors like concatenation with nil values
- "invent a document that triggers every debug case" reminds me of coverage guided fuzzing

**Comment by alerque** [comment]
It turns out I already got part way there with my last overhaul of the debug function. It already takes string or functions as arguments, which is great for speed when string conversion takes time, but it also will take a variable number of arguments and concat them. When it does so, it does so in a *NULL* safe way.

```lua
local val = nil
-- unsafe
SU.debug("foo", "string "..val)
-- safe
SU.debug("foo", "string", val)
```

I've looked through all the debug calls and converted many more of them to function callbacks and many others from string concats to additional arguments. This should improve both the speed and robustness of anything that gets close to a debug call whether the debug flag is on or off.
